extmod/modure.c

```diff
@@ -310,16 +314,21 @@ STATIC mp_obj_t re_sub_helper(size_t n_args, const mp_obj_t *args) {
         vstr_add_strn(&vstr_return, subj.begin, match->caps[0] - subj.begin);
 
         // Get replacement string
-        const char *repl = mp_obj_str_get_str((mp_obj_is_callable(replace) ? mp_call_function_1(replace, MP_OBJ_FROM_PTR(match)) : replace));
+        mp_obj_t repl_obj = (mp_obj_is_callable(replace) ? mp_call_function_1(replace, MP_OBJ_FROM_PTR(match)) : replace);
+        mp_get_buffer_raise(repl_obj, &bufinfo, MP_BUFFER_READ);
+        const char *repl = bufinfo.buf;
+        size_t repl_len = bufinfo.len;
 
         // Append replacement string to result, substituting any regex groups
-        while (*repl != '\0') {
+        while (repl_len > 0) {
```

Comment:
I think it'd be simpler (and maybe smaller code) to do:
```c
const char *repl_top = bufinfo.buf + bufinfo.len;

while (repl < repl_top) {
    ...
```

Reply:
Yes I believe you're right, not sure why I didn't do it that way from the start. I was very uncomfortable about the manual length tracking, knew right from the start it was a bad / fragile approach.